services/src/main/java/org/exoplatform/addons/gamification/storage/dao/GamificationHistoryDAO.java

```diff
@@ -450,60 +454,124 @@ public List<GamificationActionsHistory> findAllAnnouncementByChallenge(Long chal
-                                                                            boolean sortDescending,
-                                                                            int offset,
-                                                                            int limit) {
-    TypedQuery<GamificationActionsHistory> query;
-    if (sortDescending) {
-      query = getEntityManager().createNamedQuery("GamificationActionsHistory.findRealizationsByDateDescending",
-                                                  GamificationActionsHistory.class);
-    } else {
-      query = getEntityManager().createNamedQuery("GamificationActionsHistory.findRealizationsByDateAscending",
-                                                  GamificationActionsHistory.class);
-    }
-    query.setParameter(FROM_DATE_PARAM_NAME, fromDate);
-    query.setParameter(TO_DATE_PARAM_NAME, toDate);
-    query.setParameter("type", IdentityType.USER);
+  public List<GamificationActionsHistory> findRealizationsDynamicallyByFilter(RealizationsFilter filter,
+                                                                              int offset,
+                                                                              int limit) {
+    TypedQuery<GamificationActionsHistory> query = buildQueryFromFilter(filter, GamificationActionsHistory.class, false);
     if (limit > 0) {
       query.setMaxResults(limit);
     }
     if (offset > 0) {
       query.setFirstResult(offset);
     }
-    List<GamificationActionsHistory> resultList = query.getResultList();
-    return resultList == null ? Collections.emptyList() : resultList;
+    return query.getResultList();
+  }
+
+  private <T> TypedQuery<T> buildQueryFromFilter(RealizationsFilter filter, Class<T> clazz, boolean count) {
+    List<String> suffixes = new ArrayList<>();
+    List<String> predicates = new ArrayList<>();
+    buildPredicates(filter, suffixes, predicates);
+
+    TypedQuery<T> query;
+    String queryName = getQueryFilterName(suffixes, count);
+    if (filterNamedQueries.containsKey(queryName)) {
+      query = getEntityManager().createNamedQuery(queryName, clazz);
+    } else {
+      String queryContent = getQueryFilterContent(filter, predicates, count);
+      query = getEntityManager().createQuery(queryContent, clazz);
+      getEntityManager().getEntityManagerFactory().addNamedQuery(queryName, query);
+      filterNamedQueries.put(queryName, true);
+    }
+    addQueryFilterParameters(filter, query);
+    return query;
+  }
+
+  private List<String> buildPredicates(RealizationsFilter filter,
+                                       List<String> suffixes,
+                                       List<String> predicates) {
+    predicates.add("g.earnerType = :type");
+    suffixes.add(filter.getSortField());
+    suffixes.add(filter.isSortDescending() ? "Descending" : "Ascending");
+    predicates.add("g.date BETWEEN :fromDate AND :toDate");
```

Comment:
I am going also to introduce these changes concerning the out of boundaries period selector https://github.com/Meeds-io/gamification/blob/ed9f7337d0799bd56585ac2abaf10099fb06c466/services/src/main/java/org/exoplatform/addons/gamification/entities/domain/effective/GamificationActionsHistory.java#L164
@AzmiTouil

Reply:
Please avoid using `BETWEEN` to exclude all time the endDate, useful especially for pagination. By example, how to select the achievements of a mont or a week ? It should be : 
- between `2022-09-01 00:00:00` inclusive and `2022-10-01 00:00:00` exclusive
instead of
- between `2022-09-01 00:00:00` inclusive and `2022-09-30 23:59:59.999` inclusive